upgrades/upgrade_testing_framework/core/framework_state.py

```diff
@@ -13,17 +16,34 @@
 """
 class FrameworkState:
     def __init__(self, state: dict):
-        self.docker_client: DockerFrameworkClient | None = None
-        self._state_dict = state
+        self.docker_client: DockerFrameworkClient = None
+        self.source_cluster: Cluster = None
+        self.target_cluster: Cluster = None
+        self.test_config: TestConfig = None
```

Comment:
I know we discussed the `Type | None` vs `Optional[Type]` thing last time--if they're being set to `None`, should they all be `Optionals`?

Reply:
I dunno; does it matter either way?

Reply:
Ah, now I under your question.  We currently expect these to be defined over the course of a UTF run, so they're not technically "optional" in a grammatical sense.  I therefore didn't use `Optional`, and instead declared/initialized them as None.

Reply:
I don't think this is a big deal, so not worth changing in this review, but for future related code: my take would be that if there's a chance they could be None (i.e. should callers check whether it's not-None before using it?), they should be Optional. Is that fair/line up with your understanding?